Add specific message for double-encoded redirects

johngodley opened the pull request:
Some hosts seem to re-encode parameters in a http=>https redirect, which causes problems with Jetpack auth.

This checks the ‘redirect_to’ query param and if it detects a double-encoded value shows a specific error message, directing the user to a support page for further details.

This PR is an initial step to fixing the problem that is fully described in #6690:

- [x] Show a specific error related to the double-encoding problem
- [x] Link to a support page (needs writing) with further details and a list of possible solutions
- [x] Track stats / sites to see how wide-spread the problem is
- [ ] If significant enough can think about adding #6690

Note: would appreciate help regards the support page and tracking statistics!

johngodley committed:
Add specific message for double-encoded redirects

Some hosts seem to re-encode parameters in a http=>https redirect, which causes problems with Jetpack auth.

This checks the ‘redirect_to’ query param and if it detects a double-encoded value shows a specific message with link to support guide

samhotchkiss commented:
@jeherve -- are you able to help with creating this support page?

jeherve commented:
I'm happy to help, but I'm not sure what the list of possible solutions would be. @johngodley @georgeh I assume you already have some kind of predef you use for folks running into this issue with the Google Docs Add-on? Could you add the info you have to the draft I created [here](https://jetpackme.wordpress.com/wp-admin/post.php?post=16573&action=edit)?

Thanks!

johngodley commented:
Thanks @jeherve - there's no predef as such, but I've filled in a few more details. It's mostly 'fix your server', but hopefully it helps. If it looks ok to you then I think the page is good to go.

If it's possible to track stats for things like this then that would be a good addition. Is this something Jetpack can do?

If not then the PR is ready.

jeherve commented:
> If it's possible to track stats for things like this then that would be a good addition. Is this something Jetpack can do?

Yes, it can be done, through Nosara Tracks. Jetpack includes a library making it easier to track actions all across the plugin:
https://github.com/Automattic/jetpack/blob/master/_inc/lib/tracks/class.tracks-client.php

@dereksmart could probably tell you more if needed.

johngodley committed:
Track double-encode error event

Will give an indication of how widespread this is, and hopefully help with support

johngodley commented:
That's perfect, added! Event should track under `jetpack_error_double_encode`

oskosk commented:
Removing the Needs Review label as I can't get tests to run locally now. Please rebase it so it gets up to date.  And forgive me if this was a mistake on my side.

kraftbj committed:
Merge branch 'master' into add/double-encode-message

zinigor commented on class.jetpack.php:
```diff
@@ -6181,6 +6181,10 @@ function allow_wpcom_public_api_domain( $domains ) {
 
+	static function is_redirect_encoded( $redirect_url ) {
+		return preg_match( '/https?%3A%2F%2F/i', $redirect_url ) > 0;
+}
```
There seems to be a spacing problem here.

zinigor reviewed (dismissed):
Besides one minor spacing problem, and potential GDPR compliance issues I have no problems with this, thank you!

kraftbj committed:
Spacing
